### ISSUE TITLE
[Ivy] Incompatibility of Esm2015ReflectionHost with libraries compiled with NGCC

### ISSUE THREAD

**Issue by elca-lan** [angular/angular#35078]
# 🐞 bug report

### Affected Package
The issue is caused by @angular/compiler-cli

### Is this a regression?
Sort of. ~~It is a regression if you happen to upgrade from a < 3.7 to a >= 3.7 version of TypeScript~~
It did work with Angular 6 and it also worked with previous Angular 9 rc's (I'm pretty sure that rc.7 still worked).

### Description
This was _very_ tricky to track down. The problem expressed in the form that the compiler complains when running ngcc on some of the project libraries that "selector/template has to be a string". This happened for cases where we use a constant from another package as the selector or template for a @Component annotation.

A lot of debugging of the build process and some guesswork resulted in the following findings:
* In essence, the Esm2015ReflectionHost is not capable of resolving expressions of the form `MyClass.MyPublicStaticStringProperty`... in _some situations_.
* In the cases where it fails, the `reflectMember` check for `isPropertyAccess` simply does not work anymore because the syntax kind of the node is something completely unexpected.
* In `getMembersOfClass` method, the nodes of `symbol.implementation.exports` for public static properties are of syntax kind `PropertyDeclaration` when "normally" (when things work), they are of kind `PropertyAccessExpression`
* In the end, it all breaks down in the `StaticInterpreter.accessHelper` method because it ends up returning a `Reference` instead of the resolved string constant. This then triggers various extremely unhelpful downstream errors.

## 🔬 Minimal Reproduction
I don't have a minimal reproduction right now since I'm not sure how I can "build" this situation from scratch, but as soon as I find one, I will add it here. See my next issue comment for reproduction.

An "abstraction" of the code I am working with contains a component in a library in the "packages" directory
```
@Component({
    selector: SharedConstants.SecondSelector,
    template: `<div>...</div>`,
})
export class SharedDerivedComponent extends BaseComponent {
}
```
This is referencing a constant from another library in the "packages" directory
```
export declare class SharedConstants {
    static readonly SecondSelector = "lib-shared-derived";
}
```

## 🔥 Exception or Error
Some of the messages that result from this are:
<pre><code>
NG1010: selector must be a string
</code></pre>

<pre><code>
NG1010: template must be a string
</code></pre>

<pre><code>
NG1006: Two incompatible decorators on class
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.11
Node: 10.16.0
OS: win32 x64

Angular: 9.0.0-rc.11
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.0-rc.11
@angular-devkit/build-angular      0.900.0-rc.11
@angular-devkit/build-ng-packagr   0.900.0-rc.11
@angular-devkit/build-optimizer    0.900.0-rc.11
@angular-devkit/build-webpack      0.900.0-rc.11
@angular-devkit/core               9.0.0-rc.11
@angular-devkit/schematics         9.0.0-rc.11
@ngtools/webpack                   9.0.0-rc.11
@schematics/angular                9.0.0-rc.11
@schematics/update                 0.900.0-rc.11
ng-packagr                         9.0.0-rc.7
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.8.3
</code></pre>

**Comment by elca-lan** [comment]
Somehow it does **not** seem to be related to the TypeScript version even though that is the last thing I changed before everything broke...

I have however created a sample that at least demonstrates the actual problem that I am experiencing: https://github.com/elca-lan/angular9-lib-constant-selectors

After checkout, just run the following commands and this will result in some of the described compiler errors
```
npm ci
ng build shared --prod
ng build business --prod
ng serve
```
The project consists of three "layers": the **shared** library, the **business** library (which depends on the shared library) and the main application. In all three layers, I use constants from the shared library as component selectors:
* `SharedComponent` uses `Constants.SharedSelector`
* `BusinessComponent` uses `Constants.BusinessSelector`
* `AppComponent` uses `Constants.AppSelector`

Now, if we _change_ the selector of the `BusinessComponent` (see comments in that class) to something else like a constant on the component class itself, and re-build the libraries and the project, then it suddenly works. So.... only the "intermediate" business library is not allowed to reference constants from shared and everywhere else, these selectors work.

There is definitely something wrong here. And it _is_ related to the `Esm2015ReflectionHost`, `reflectMember` and `getMembersOfClass`, but it does not seem to be specific to a TS version (I tested the project with a downgraded TS version and it still behaves the same).

**Comment by IgorMinar** [comment]
thanks for the repro repo, we'll investigate next week.

**Comment by elca-lan** [comment]
Some more detailed information can't hurt.

The class containing the constants in the sample project is compiled to the following fesm2015 code

```
class Constants {
}
Constants.SharedSelector = 'lib-shared';
Constants.BusinessSelector = 'lib-business';
Constants.AppSelector = 'app-root';
```

When the partial evaluator is trying to evaluate the expression accessing the constant value, the `Esm2015ReflectionHost.getMembersOfSymbol` method finds the relevant node(s) in `symbol.implementation.exports` (**not** in `symbol.declaration.exports` like it does for variable declarations where everything works fine as far as I can tell).

The `Esm2015ReflectionHost.reflectMember` is then receiving the following node object

```
NodeObject {
  pos: 84,
  end: 140,
  flags: 8388608,
  modifierFlagsCache: 536871008,
  transformFlags: 0,
  parent:
   NodeObject {
     pos: 0,
     end: 190,
     flags: 8388608,
     modifierFlagsCache: 536870915,
     transformFlags: 0,
     parent:
      SourceFileObject {
        pos: 0,
        end: 192,
        flags: 8388672,
        modifierFlagsCache: 0,
        transformFlags: 0,
        parent: undefined,
        kind: 288,
        text:
         'export declare class Constants {\r\n    static readonly SharedSelector = "lib-shared";\r\n    static readonly BusinessSelector = "lib-business";\r\n    static readonly AppSelector = "app-root";\r\n}\r\n',
        bindDiagnostics: [],
        bindSuggestionDiagnostics: undefined,
        languageVersion: undefined,
        fileName:
         'H:/development/angular9-lib-constant-selectors/dist/shared/lib/constants.d.ts',
        languageVariant: 0,
        isDeclarationFile: true,
        scriptKind: 3,
        pragmas: Map {},
        checkJsDirective: undefined,
        referencedFiles: [],
        typeReferenceDirectives: [],
        libReferenceDirectives: [],
        amdDependencies: [],
        hasNoDefaultLib: false,
        statements: [Array],
        endOfFileToken: [TokenObject],
        externalModuleIndicator: [Circular],
        nodeCount: 23,
        identifierCount: 4,
        identifiers: [Map],
        parseDiagnostics: [],
        path:
         'H:/development/angular9-lib-constant-selectors/dist/shared/lib/constants.d.ts',
        resolvedPath:
         'H:/development/angular9-lib-constant-selectors/dist/shared/lib/constants.d.ts',
        originalFileName:
         'H:/development/angular9-lib-constant-selectors/dist/shared/lib/constants.d.ts',
        imports: [],
        moduleAugmentations: [],
        ambientModuleNames: [],
        resolvedModules: undefined,
        symbol: [SymbolObject],
        locals: [Map],
        nextContainer: [Circular],
        symbolCount: 7,
        classifiableNames: [Map] },
     kind: 244,
     decorators: undefined,
     modifiers: [ [TokenObject], [TokenObject], pos: 0, end: 14 ],
     name:
      IdentifierObject {
        pos: 20,
        end: 30,
        flags: 8388608,
        modifierFlagsCache: 0,
        transformFlags: 0,
        parent: [Circular],
        kind: 75,
        escapedText: 'Constants',
        flowNode: [Object] },
     typeParameters: undefined,
     heritageClauses: undefined,
     members:
      [ [NodeObject], [Circular], [NodeObject], pos: 32, end: 187 ],
     symbol:
      SymbolObject {
        flags: 32,
        escapedName: 'Constants',
        declarations: [Array],
        exports: [Map],
        members: Map {},
        valueDeclaration: [Circular],
        parent: [SymbolObject] },
     localSymbol:
      SymbolObject {
        flags: 1048576,
        escapedName: 'Constants',
        declarations: [Array],
        parent: undefined,
        exportSymbol: [SymbolObject] } },
  kind: 158,
  decorators: undefined,
  modifiers:
   [ TokenObject {
       pos: 84,
       end: 96,
       flags: 8388608,
       modifierFlagsCache: 0,
       transformFlags: 0,
       parent: [Circular],
       kind: 119 },
     TokenObject {
       pos: 96,
       end: 105,
       flags: 8388608,
       modifierFlagsCache: 0,
       transformFlags: 0,
       parent: [Circular],
       kind: 137 },
     pos: 84,
     end: 105 ],
  name:
   IdentifierObject {
     pos: 105,
     end: 122,
     flags: 8388608,
     modifierFlagsCache: 0,
     transformFlags: 0,
     parent: [Circular],
     kind: 75,
     escapedText: 'BusinessSelector',
     flowNode: { flags: 2 } },
  questionToken: undefined,
  type: undefined,
  initializer:
   TokenObject {
     pos: 124,
     end: 139,
     flags: 8388608,
     modifierFlagsCache: 0,
     transformFlags: 0,
     parent: [Circular],
     kind: 10,
     text: 'lib-business' },
  symbol:
   SymbolObject {
     flags: 4,
     escapedName: 'BusinessSelector',
     declarations: [ [Circular] ],
     valueDeclaration: [Circular],
     parent:
      SymbolObject {
        flags: 32,
        escapedName: 'Constants',
        declarations: [Array],
        exports: [Map],
        members: Map {},
        valueDeclaration: [NodeObject],
        parent: [SymbolObject] } } }
```

As you can see, the node is of syntax kind `PropertyDeclaration` (158 in TS 3.7.5) which ends up _not_ assigning a `value` to the returned `ClassMember`. Instead it has an `initializer` token containing the property value. This case is simply not covered by the ESM2015 host.

The analogous `TypeScriptReflectionHost._reflectMember` method explicitly deals with this case with
```
if (ts.isPropertyDeclaration(node)) {
    kind = host_1.ClassMemberKind.Property;
    value = node.initializer || null;
}
```
which assigns the initializer token to the `ClassMember` value which seems to me is what _should_ happen in the ESM2015 host as well.

**Comment by JoostK** [comment]
Hi @elca-lan, you mention the following:

> This is referencing a constant from another library in the "packages" directory

This is where the root of your issue is, as the Ivy compiler can no longer read values defined in external compilations. The reason is that it would break locality, where an "internal" change in a library (e.g. changing the value of a selector constant) would require recompiling an upstream consumer.

This is mentioned in the [Ivy compatibility guide](https://v9.angular.io/guide/ivy-compatibility):

> Foreign functions or foreign constants in decorator metadata aren't statically resolvable (previously, you could import a constant or function from another compilation unit, like a library, and use that constant/function in your @NgModule definition).

I notice that the `.d.ts` that is published as part of the library _does_ in fact include the values, so we need determine if we should support this setup.

**Comment by elca-lan** [comment]
Thank you for your reply @JoostK, but I think I'll have to argue with you on that one.

If that statement from the documentation were strictly true, then things like the `forRoot(...)`
providers like it is used for the [RouterModule](https://next.angular.io/api/router/RouterModule) wouldn't work, right? `forRoot` is a function from a "foreign" package which is used in the decorator metadata of any Angular application that uses routing...

Also, if that statement were true, that would be _extremely_ limiting and make the whole Angular framework a lot less powerful.

From a practical point of view, if that was not supported, then the compiler should raise an error
- preferrably with a reasonable error message - if it encounters anything like that and not just the NGCC.

It's not the Ivy compiler that fails with this code. It is the NGCC backwards compatibility compiler. If I switch the libraries to also `enableIvy` in production builds, then it _works_.

Other things that work with the libraries back to non-ivy builds are:
* Referencing an `export const VarBusinessSelector = 'lib-business';` defined in the sahred library
* Referencing a function `export function FuncBusinessSelector(): string { return 'lib-business'; }` defined in the shared library

It would be quite sneaky to declare those features that clearly _work_ as "not supported" as people generally don't consult the documentation to check if something is "legal" if it works and makes sense.

As you correctly pointed out, the class containing the constants is of course part of the public API of that library and I don't see how that would violate locality.

In essence, I think that the statement from the upgrade instructions is wildly inaccurate or at the very least misleading. It would be great if you could provide some clarification with regards to what this actually _means_ since I don't think that `RouterModule.forRoot(ROUTES)` is going to change. I would also like to know what definition of "locality" you are referring to, since it does not seem to match any that I'm aware of.

**Comment by petebacondarwin** [comment]
@JoostK - if we are able to get "constant" values from the typings, then I think it should be supported (at least via ngcc even if we do not support it going forward for ngtsc).

**Comment by JoostK** [comment]
@elca-lan Please find my comments inline: 
 
> If that statement from the documentation were strictly true, then things like the `forRoot(...)`
> providers like it is used for the [RouterModule](https://next.angular.io/api/router/RouterModule) wouldn't work, right? `forRoot` is a function from a "foreign" package which is used in the decorator metadata of any Angular application that uses routing...

The `forRoot()` idiom continues to work due to a change in the `ModuleWithProviders` type that is returned by such functions, where a generic type parameter has been added that represents the type of the `NgModule` that is imported. Since this type ends up in the typings, it is no longer "internal".

> Also, if that statement were true, that would be _extremely_ limiting and make the whole Angular framework a lot less powerful.

In your case, the TypeScript compiler _did_ include the value assignment in the typings as it's assigned into a readonly property. That makes the value available to inspect _without_ looking into the actual source code, which is something that can therefore be supported (and ngtsc in fact does, as you found out). From that perspective, it is indeed a bug that ngcc does not support this use case.
 
> From a practical point of view, if that was not supported, then the compiler should raise an error
> 
> * preferrably with a reasonable error message - if it encounters anything like that and not just the NGCC.

Absolutely, agreed!

> It's not the Ivy compiler that fails with this code. It is the NGCC backwards compatibility compiler. If I switch the libraries to also `enableIvy` in production builds, then it _works_.
> 
> Other things that work with the libraries back to non-ivy builds are:
> 
> * Referencing an `export const VarBusinessSelector = 'lib-business';` defined in the sahred library

I am not certain about how TypeScript would represent a const in the typings, but I suspect it would be widened to `string` instead of the literal type `lib-business`. Therefore, the actual value would have been lost in the typings. I think this can be forced using `export const VarBusinessSelector = 'lib-business' as const;` as that would keep the literal type.

Your example with `static readonly` is apparently kept as literal type without the `as const` cast, I'd have to check how a `const` variable is transformed into the typings file.

> * Referencing a function `export function FuncBusinessSelector(): string { return 'lib-business'; }` defined in the shared library

This will no longer work, as the `"lib-business"` value will not be available in the typings. In ViewEngine, this did work as the value side of a library was captured in the `metadata.json` file format when compiling the library. Such metadata is no longer captured,

> It would be quite sneaky to declare those features that clearly _work_ as "not supported" as people generally don't consult the documentation to check if something is "legal" if it works and makes sense.

Agreed, the compiler should be able to provide useful and actionable diagnostics for situations that are not supported.

> As you correctly pointed out, the class containing the constants is of course part of the public API of that library and I don't see how that would violate locality.
> 
> In essence, I think that the statement from the upgrade instructions is wildly inaccurate or at the very least misleading. It would be great if you could provide some clarification with regards to what this actually _means_ since I don't think that `RouterModule.forRoot(ROUTES)` is going to change. I would also like to know what definition of "locality" you are referring to, since it does not seem to match any that I'm aware of.

In essence, values that are only available inside the source code but not the typings are "internal" and not available to the ngtsc compiler. My usage of the term _locality_ may not be very accurate, as it has various associations (where my usage does not correspond with how locality is typically referred to)


I think there's three action items here:
1. Fix ngcc to detect .d.ts files and fall back into the `TypeScriptReflectionHost` host.
2. Document the exact patterns that are, and that are not supported. The Ivy compatibility guide may need some clarification in this area as well.
3. Work towards better error messages in the compiler, with actionable suggestions. With Ivy, one of the goals is _much_ improved error messages and this is one area where improvements are to be made.

**Comment by elca-lan** [comment]
@JoostK thanks **a lot** for the very detailed explanation and action items. With all the backwards compatibility going on, it is sometimes hard to tell what the actual design principles behind Ivy are, but this cleared things up big time for me.

It seems from your description that one of the goals of Ivy is to get away from the parsing of source files and instead mostly rely on typing information. This is something I can totally get on board with and it also gives me a concrete idea of what to look out for to avoid issues in the future.

I will also check the `ModuleWithProviders` to make sure my code base is compatible with the latest patterns in providers.

I guess it's time to look into the contribution process for Angular in more detail :)

**Comment by JoostK** [comment]
I opened #36089 to fix the issue, and #36092 to better document this. That covers the first two of the proposed action items, the last one should be tackled in a larger effort around error reporting.
